diag_manager/fms_diag_dlinked_list.F90

```diff
@@ -0,0 +1,323 @@
+!* the terms of the GNU Lesser General Public License as published by
+!* the Free Software Foundation, either version 3 of the License, or (at
+!* your option) any later version.
+!*
+!* FMS is distributed in the hope that it will be useful, but WITHOUT
+!* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+!* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+!* for more details.
+!*
+!* You should have received a copy of the GNU Lesser General Public
+!* License along with FMS.  If not, see <http://www.gnu.org/licenses/>.
+!***********************************************************************
+
+!> @defgroup fms_diag_dlinked_list_mod fms_diag_dlinked_list_mod
+!> @ingroup diag_manager
+!> @brief fms_diag_dlinked_list_mod defines a generic doubly linked
+!! list class and an iterator class for traversing the list.
+!!
+!> @author Miguel Zuniga
+!!
+!! <TT>fms_diag_dlinked_list_mod</TT>  defines a generic doubly linked
+!! list class and an iterator class for traversing the list. It is
+!! generic in the sense that the elements or objects it contains are
+!! "class(*)" objects.  If additional typecheking or psossibly a
+!! slightly different user interface is desired, consider creating
+!! a wrapper or another class with this one for a memeber element and
+!! procedures that are trivially implemeted by using this class.
+!!
+!! This version is roughly a fortran translation of the C++ doubly linked list
+!! class in the book ``Data Structures And Algorithm Analysis in C++", 3rd Edition,
+!! by Mark Allen Weiss.
+
+!> @file
+!> @brief File for @ref fms_diag_dlinked_list_mod
+!> @addtogroup fms_diag_dlinked_list_mod
+!> @{
+MODULE fms_diag_dlinked_list_mod
+  USE fms_mod, ONLY: error_mesg, FATAL, WARNING, NOTE
+  implicit none
+  !!TODO: COnsider setting the access (public,private) to functions, etc.
+  !> Linked doubly-linked list node type.
+  type, public:: FmsDlListNode_t
+    private
+    class(*), pointer :: data => null()        !< The data pointed to by the node.
+    type(FmsDlListNode_t), pointer :: next => null() !< A pointer to the previous node.
+    type(FmsDlListNode_t), pointer :: prev => null() !< A pointer to the next node.
+  end type FmsDlListNode_t
+
+  !> Linked list iterator
+  type, public :: FmsDllIterator_t
+  private
+    type(FmsDlListNode_t), pointer :: current  !< A pointer to the current node.
+    type(FmsDlListNode_t), pointer :: end      !< A sentinel (non-data) node.
+  contains
+    procedure :: has_data => literator_has_data !< Function returns true is there is data in the iterator.
+    procedure :: next => literator_next !< Function moves the iterator to the next data element.
+    procedure :: get => literator_data !< Function return a pointer to the current data.
+    procedure :: get_current_node_pointer => get_current_node_ptr !< Return  the current node pointer.
+  end type FmsDllIterator_t
+
+
+  type, public :: FmsDlList_t
```

Comment:
Needs a doxygen description for this to be parsed, I woulld include the two comments below this in the description as well (second one needs another `!`).

Reply:
Done.